# 05-5. 내장 함수
파이썬 내장 함수는 외부 모듈과 달리 import가 필요하지 않기 때문에 아무런 설정 없이 바로 사용할 수 있다.

### abs(x) 
: 어떤 숫자를 입력받았을 때, 그 숫자의 절댓값을 돌려주는 함수

In [2]:
print(abs(-3))
print(abs(-1.2))

3
1.2


### all(x) 
: 반복 가능한(iterable) 자료형 x를 입력 인수로 받으며 이 x가 모두 참이면 True, 거짓이 하나라도 있으면 False를 돌려준다.

In [3]:
all([1, 2, 3])

True

In [4]:
all([1, 2, 3, 0])

False

리스트 자료형 [1, 2, 3, 0] 중에서 요소 0은 거짓이므로 False를 돌려준다.

### any
: x 중 하나라도 참이 있으면 True를 돌려주고, x가 모두 거짓일 때에만 False를 돌려준다. 

In [5]:
any([1, 2, 3, 0])

True

In [6]:
any([0, ""])

False

리스트 자료형 [0, ""]의 요소 0과 ""은 모두 거짓이므로 False를 돌려준다.

### chr
: 아스키(ASCII) 코드 값을 입력받아 그 코드에 해당하는 문자를 출력하는 함수이다.

In [7]:
chr(97)

'a'

In [8]:
chr(48)

'0'

### dir
: 객체가 자체적으로 가지고 있는 변수나 함수를 보여 준다.

In [10]:
print(dir([1, 2, 3])) #리스트 객체 관련 메서드

['__add__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__iadd__', '__imul__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__rmul__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']


In [12]:
print(dir({'1':'a'})) #딕셔너리 객체 관련 메서드

['__class__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'clear', 'copy', 'fromkeys', 'get', 'items', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values']


### divmod(a,b)
: 2개의 숫자를 입력으로 받는다. 그리고 a를 b로 나눈 몫과 나머지를 튜플 형태로 돌려주는 함수이다.

In [13]:
divmod(7, 3)

(2, 1)

### enumerate
: 순서가 있는 자료형(리스트, 튜플, 문자열)을 입력으로 받아 인덱스 값을 포함하는 enumerate 객체를 돌려준다.

In [15]:
print(enumerate(['body', 'foo', 'bar']))

'body', 'foo', 'bar'이 각각의 인덱스 값과 매치된 enumerate 객체가 생성된다.

In [16]:
for i, name in enumerate(['body', 'foo', 'bar']):
    print(i, name)  #i가 인덱스를 형성

0 body
1 foo
2 bar


순서 값과 함께 body, foo, bar가 순서대로 출력되었다.

for문처럼 반복되는 구간에서 객체가 현재 어느 위치에 있는지 알려 주는 인덱스 값이 필요할 대 사용하면 유용하다.

### eval(expression)
:실행 가능한 문자열(1+2, 'hi' + 'a' 같은 것)을 입력으로 받아 문자열을 실행한 결괏값을 돌려주는 함수이다.

In [17]:
eval('1+2')

3

In [18]:
eval("'hi' + 'a'")

'hia'

### filter
: 첫 번째 인수로 함수 이름을, 두 번째 인수로 그 함수에 차례로 들어갈 반복 가능한 자료형을 받는다. 그리고 두 번째 인수인 자료형의 요소를 첫 번째 인수로 입력한 함수에 넣어 참인 것만 돌려준다.

In [19]:
def positive(x):
    return x > 0

print(list(filter(positive, [1, -3, 2, 0, -5, 6])))

[1, 2, 6]


### hex(x)
: 정수 값을 입력받아 16진수(hexadecimal)로 변환하여 돌려주는 함수이다.

In [20]:
hex(234)

'0xea'

In [21]:
hex(3)

'0x3'

### id(object)
: 객체를 입력받아 객체의 고유 주소 값(레퍼런스)을 돌려주는 함수이다.

In [25]:
a = 3
print(id(3))
print(id(a))

b = a
print(id(b))

140734048015232
140734048015232
140734048015232
140734048015264


 3, a, b는 고유 주소 값이 모두 140734048015232이다. 즉 3, a, b가 모두 같은 객체를 가리키고 있다.

In [26]:
print(id(4))

140734048015264


만약 id(4)라고 입력하면 4는 3, a, b와 다른 객체이므로 다른 고유 주소 값이 출력된다.

### input
: 사용자 입력을 받는 함수이다. 매개변수로 문자열을 주면 그 문자열은 프롬프트가 된다.

In [29]:
a = input()

hi


위 코드를 실행하면 문자열을 입력할 수 있는 칸이 생성된다. 그 칸에 'hi'를 입력했다.

In [30]:
a

'hi'

a에 사용자가 입력한 'hi'가 저장된다.

In [31]:
b = input("Enter: ")

Enter: hi


칸 옆에 적절한 문구를 설정할 수 있다.

### int(x)
: 문자열 형태의 숫자나 소수점이 있는 숫자 등을 정수 형태로 돌려주는 함수이다.

In [32]:
int('3')

3

In [33]:
int(3.4)

3

int(x, radix)는 radix 진수로 표현된 문자열 x를 10진수로 변환하여 돌려준다.

In [34]:
int('11', 2) #2진수로 표현된 문자열

3

In [35]:
int('1A', 16) #16진수로 표현된 문자열

26

### isinstance(object, class )
: 첫 번째 인수로 인스턴스, 두 번째 인수로 클래스 이름을 받는다. 입력으로 받은 인스턴스가 그 클래스의 인스턴스인지를 판단하여 참이면 True, 거짓이면 False를 돌려준다.

In [36]:
class Person: pass #아무 기능이 없는 Person이라는 클래스 생성

In [39]:
a = Person() #Person 클래스의 인스턴스 a 생성
isinstance(a, Person) #a가 Person의 인스턴스인지 확인

True

In [40]:
b = 3
isinstance(b, Person)

False

b는 Person의 인스턴스가 아니므로 False반환

### len(s)
: 입력값 s의 길이(요소의 전체 개수)를 돌려주는 함수이다

In [41]:
len("python")

6

In [42]:
len([1,2,3])

3

### list(s)
: 반복 가능한 자료형 s를 입력받아 리스트로 만들어 돌려주는 함수이다.

In [43]:
list("python")  

['p', 'y', 't', 'h', 'o', 'n']

In [44]:
list((1,2,3))

[1, 2, 3]

### map(f, iterable)
: 함수(f)와 반복 가능한(iterable) 자료형을 입력으로 받는다. map은 입력받은 자료형의 각 요소를 함수 f가 수행한 결과를 묶어서 돌려주는 함수이다.

In [48]:
def two_times(numberList):
    result = [ ]
    for number in numberList:
        result.append(number*2)
    return result

result = two_times([1, 2, 3, 4])
print(result)

[2, 4, 6, 8]


two_times 함수는 리스트 요소를 입력받아 각 요소에 2를 곱한 결괏값을 돌려준다. 위 예제는 map 함수를 사용하면 다음처럼 바꿀 수 있다. 

In [47]:
def two_times(x): 
    return x*2

list(map(two_times, [1, 2, 3, 4]))

[2, 4, 6, 8]

### max
: 반복 가능한 자료형을 입력받아 그 최댓값을 돌려주는 함수이다.

In [49]:
max([1, 2, 3])

3

In [50]:
max("python")

'y'

### min
: 반복 가능한 자료형을 입력받아 그 최솟값을 돌려주는 함수이다.

In [51]:
min([1, 2, 3])

1

In [52]:
min("python")

'h'

### open(filename, [mode])
: "파일 이름"과 "읽기 방법"을 입력받아 파일 객체를 돌려주는 함수이다. 읽기 방법(mode)을 생략하면 기본값인 읽기 전용 모드(r)로 파일 객체를 만들어 돌려준다.

- w : 쓰기 모드로 파일 열기
- r : 읽기 모드로 파일 열기
- a : 추가 모드로 파일 열기
- b : 바이너리 모드로 파일 열기.
b는 w,r,a와 함께 사용된다.

In [2]:
fread = open("read_mode.txt", 'r')
print(fread) #read_mode.txt 객체가 생성된다.

<_io.TextIOWrapper name='read_mode.txt' mode='r' encoding='cp949'>


### ord(c)
: 문자의 아스키 코드 값을 돌려주는 함수이다. chr 함수와 반대이다.

In [3]:
ord('a')

97

In [4]:
ord('0')

48

### pow(x, y)
: x의 y 제곱한 결괏값을 돌려주는 함수이다.

In [5]:
pow(2, 4)

16

### range(start, stop)
: 입력받은 숫자에 해당하는 범위 값을 반복 가능한 객체로 만들어 돌려준다. for문과 함께 자주 사용한다.

In [6]:
list(range(5))

[0, 1, 2, 3, 4]

시작 숫자를 지정해 주지 않으면 range 함수는 0부터 시작한다.

In [7]:
list(range(5, 10))

[5, 6, 7, 8, 9]

입력으로 주어지는 2개의 인수는 시작 숫자와 끝 숫자를 나타낸다. 단 끝 숫자는 해당 범위에 포함되지 않는다는 것에 주의하자.

In [8]:
list(range(1, 10, 2))

[5, 6, 7, 8, 9]

세 번째 인수는 숫자 사이의 거리를 말한다.

### round(number[, ndigits]) 
: 숫자를 입력받아 반올림해 주는 함수이다.

In [11]:
round(5.678, 2)

5.68

round 함수의 두 번째 매개변수는 반올림하여 표시하고 싶은 소수점의 자릿수(ndigits)이다.

In [10]:
round(4.06)

4

ndigits를 설정하지 않으면 일의 자리까지 반올림한다.

### sorted(iterable)
: 입력값을 정렬한 후 그 결과를 리스트로 돌려주는 함수이다.

In [12]:
sorted([3, 1, 2])

[1, 2, 3]

In [13]:
sorted(['a', 'c', 'b'])

['a', 'b', 'c']

In [15]:
sorted("zero")

['e', 'o', 'r', 'z']

### str(object)
: 문자열 형태로 객체를 변환하여 돌려주는 함수이다.

In [16]:
str(3)

'3'

In [17]:
str('hi')

'hi'

In [18]:
str('hi'.upper()) #upper는 문자열을 대문자로 바꿔주는 메서드

'HI'

### sum(iterable)
: 입력받은 리스트나 튜플의 모든 요소의 합을 돌려주는 함수이다.

In [19]:
sum([1,2,3])

6

In [20]:
sum((4,5,6))

15

### tuple(iterable)
: 반복 가능한 자료형을 입력받아 튜플 형태로 바꾸어 돌려주는 함수이다.

In [21]:
tuple("abc")

('a', 'b', 'c')

In [22]:
tuple([1, 2, 3])

(1, 2, 3)

### type(object)
: 입력값의 자료형이 무엇인지 알려 주는 함수이다.

In [24]:
type("abc")

str

In [25]:
type([ ])

list

In [23]:
type(open("read_mode.txt", 'w'))

_io.TextIOWrapper

### zip(iterable)
: 동일한 개수로 이루어진 자료형을 같은 인덱스끼리 묶어 주는 역할을 하는 함수이다.

In [26]:
list(zip([1, 2, 3], [4, 5, 6]))

[(1, 4), (2, 5), (3, 6)]

In [27]:
list(zip("abc", "def"))

[('a', 'd'), ('b', 'e'), ('c', 'f')]